File: ecommerce_integrations/shopify/page/shopify_import_products/shopify_import_products.js
frappe.provide("shopify");

frappe.pages["shopify-import-products"].on_page_load = function (wrapper) {
	let page = frappe.ui.make_app_page({
		parent: wrapper,
		title: "Import Shopify Products",
		single_column: true,
	});
	// eslint-disable-next-line no-undef
	new shopify.ProductImporter(wrapper);
};

// eslint-disable-next-line no-undef
shopify.ProductImporter = class {
	constructor(wrapper) {
		this.wrapper = $(wrapper).find(".layout-main-section");
		this.page = wrapper.page;
		this.init();
		this.syncRunning = false;
	}

	init() {
		frappe.run_serially([
			() => this.addMarkup(),
			() => this.fetchProductCount(),
			() => this.addTable(),
			() => this.checkSyncStatus(),
			() => this.listen(),
		]);
	}

	async checkSyncStatus() {
		const jobs = await frappe.db.get_list("RQ Job", {
			filters: { status: ("in", ("queued", "started")) },
		});
		this.syncRunning =
			jobs.find(
				(job) => job.job_name == "shopify.job.sync.all.products"
			) !== undefined;

		if (this.syncRunning) {
			this.toggleSyncAllButton();
			this.logSync();
		}
	}

	addMarkup() {
		const _markup = $(`
            <div class="row">
                <div class="col-lg-8 d-flex align-items-stretch">
                    <div class="card border-0 shadow-sm p-3 mb-3 w-100 rounded-sm" style="background-color: var(--card-bg)">
                        <h5 class="border-bottom pb-2">Products in Shopify</h5>
                        <div id="shopify-product-list">
                            <div class="text-center">Loading...</div>
                        </div>
                        <div class="shopify-datatable-footer mt-2 pt-3 pb-2 border-top text-right" style="display: none">
                            <div class="btn-group">
                                <button type="button" class="btn btn-sm btn-default btn-paginate btn-prev">Prev</button>
                                <button type="button" class="btn btn-sm btn-default btn-paginate btn-next">Next</button>
                            </div>
                        </div>
                    </div>
                </div>
                <div class="col-lg-4 d-flex align-items-stretch">
                    <div class="w-100">
                        <div class="card border-0 shadow-sm p-3 mb-3 rounded-sm" style="background-color: var(--card-bg)">
                            <h5 class="border-bottom pb-2">Synchronization Details</h5>
                            <div id="shopify-sync-info">
                                <div class="py-3 border-bottom">
                                    <button type="button" id="btn-sync-all" class="btn btn-xl btn-primary w-100 font-weight-bold py-3">Sync all Products</button>
                                </div>
                                <div class="product-count py-3 d-flex justify-content-stretch">
                                    <div class="text-center p-3 mx-2 rounded w-100" style="background-color: var(--bg-color)">
                                        <h2 id="count-products-shopify">-</h2>
                                        <p class="text-muted m-0">in Shopify</p>
                                    </div>
                                    <div class="text-center p-3 mx-2 rounded w-100" style="background-color: var(--bg-color)">
                                        <h2 id="count-products-erpnext">-</h2>
                                        <p class="text-muted m-0">in ERPNext</p>
                                    </div>
                                    <div class="text-center p-3 mx-2 rounded w-100" style="background-color: var(--bg-color)">
                                        <h2 id="count-products-synced">-</h2>
                                        <p class="text-muted m-0">Synced</p>
                                    </div>
                                </div>
                            </div>
                        </div>

                        <div class="card border-0 shadow-sm p-3 mb-3 rounded-sm" style="background-color: var(--card-bg); display: none;">
                            <h5 class="border-bottom pb-2">Sync Log</h5>
                            <div class="control-value like-disabled-input for-description overflow-auto" id="shopify-sync-log" style="max-height: 500px;"></div>
                        </div>

                    </div>
                </div>
            </div>
        `);

		this.wrapper.append(_markup);
	}

	async fetchProductCount() {
		try {
			const {
				message: { erpnextCount, shopifyCount, syncedCount },
			} = await frappe.call({
				method: "ecommerce_integrations.shopify.page.shopify_import_products.shopify_import_products.get_product_count",
			});

			this.wrapper.find("#count-products-shopify").text(shopifyCount);
			this.wrapper.find("#count-products-erpnext").text(erpnextCount);
			this.wrapper.find("#count-products-synced").text(syncedCount);
		} catch (error) {
			frappe.throw(__("Error fetching product count."));
		}
	}

	async addTable() {
		const listElement = this.wrapper.find("#shopify-product-list")[0];
		this.shopifyProductTable = new frappe.DataTable(listElement, {
			columns: [
				// {
				//     name: 'Image',
				//     align: 'center',
				// },
				{
					name: "ID",
					align: "left",
					editable: false,
					focusable: false,
				},
				{
					name: "Name",
					editable: false,
					focusable: false,
				},
				{
					name: "SKUs",
					editable: false,
					focusable: false,
				},
				{
					name: "Status",
					align: "center",
					editable: false,
					focusable: false,
				},
				{
					name: "Action",
					align: "center",
					editable: false,
					focusable: false,
				},
			],
			data: await this.fetchShopifyProducts(),
			layout: "fixed",
		});

		this.wrapper.find(".shopify-datatable-footer").show();
	}

	async fetchShopifyProducts(from_ = null) {
		try {
			const {
				message: { products, nextUrl, prevUrl },
			} = await frappe.call({
				method: "ecommerce_integrations.shopify.page.shopify_import_products.shopify_import_products.get_shopify_products",
				args: { from_ },
			});
			this.nextUrl = nextUrl;
			this.prevUrl = prevUrl;

			const shopifyProducts = products.map((product) => ({
				// 'Image': product.image && product.image.src && `<img style="height: 50px" src="${product.image.src}">`,
				ID: product.id,
				Name: product.title,
				SKUs:
					product.variants &&
					product.variants.map((a) => `${a.sku}`).join(", "),
				Status: this.getProductSyncStatus(product.synced),
				Action: !product.synced
					? `<button type="button" class="btn btn-default btn-xs btn-sync mx-2" data-product="${product.id}"> Sync </button>`
					: `<button type="button" class="btn btn-default btn-xs btn-resync mx-2" data-product="${product.id}"> Re-sync </button>`,
			}));

			return shopifyProducts;
		} catch (error) {
			frappe.throw(__("Error fetching products."));
		}
	}

	getProductSyncStatus(status) {
		return status
			? `<span class="indicator-pill green">Synced</span>`
			: `<span class="indicator-pill orange">Not Synced</span>`;
	}

	listen() {
		// sync a product from table
		this.wrapper.on("click", ".btn-sync", (e) => {
			const _this = $(e.currentTarget);

			_this.prop("disabled", true).text("Syncing...");

			const product = _this.attr("data-product");
			this.syncProduct(product).then((status) => {
				if (!status) {
					frappe.throw(__("Error syncing product"));
					_this.prop("disabled", false).text("Sync");
					return;
				}

				_this
					.parents(".dt-row")
					.find(".indicator-pill")
					.replaceWith(this.getProductSyncStatus(true));

				_this.replaceWith(
					`<button type="button" class="btn btn-default btn-xs btn-resync mx-2" data-product="${product}"> Re-sync </button>`
				);
			});
		});

		this.wrapper.on("click", ".btn-resync", (e) => {
			const _this = $(e.currentTarget);

			_this.prop("disabled", true).text("Syncing...");

			const product = _this.attr("data-product");
			this.resyncProduct(product)
				.then((status) => {
					if (!status) {
						frappe.throw(__("Error syncing product"));
						return;
					}

					_this
						.parents(".dt-row")
						.find(".indicator-pill")
						.replaceWith(this.getProductSyncStatus(true));

					_this.prop("disabled", false).text("Re-sync");
				})
				.catch((ex) => {
					_this.prop("disabled", false).text("Re-sync");
					frappe.throw(__("Error syncing Product"));
				});
		});

		// pagination
		this.wrapper.on("click", ".btn-prev,.btn-next", (e) =>
			this.switchPage(e)
		);

		// sync all products
		this.wrapper.on("click", "#btn-sync-all", (e) => this.syncAll(e));
	}

	async syncProduct(product) {
		const { message: status } = await frappe.call({
			method: "ecommerce_integrations.shopify.page.shopify_import_products.shopify_import_products.sync_product",
			args: { product },
		});

		if (status) this.fetchProductCount();

		return status;
	}

	async resyncProduct(product) {
		const { message: status } = await frappe.call({
			method: "ecommerce_integrations.shopify.page.shopify_import_products.shopify_import_products.resync_product",
			args: { product },
		});

		if (status) this.fetchProductCount();

		return status;
	}

	async switchPage({ currentTarget }) {
		const _this = $(currentTarget);

		$(".btn-paginate").prop("disabled", true);
		this.shopifyProductTable.showToastMessage("Loading...");

		const newProducts = await this.fetchShopifyProducts(
			_this.hasClass("btn-next") ? this.nextUrl : this.prevUrl
		);

		this.shopifyProductTable.refresh(newProducts);

		$(".btn-paginate").prop("disabled", false);
		this.shopifyProductTable.clearToastMessage();
	}

	syncAll() {
		this.checkSyncStatus();
		this.toggleSyncAllButton();

		if (this.syncRunning) {
			frappe.msgprint(__("Sync already in progress"));
		} else {
			frappe.call({
				method: "ecommerce_integrations.shopify.page.shopify_import_products.shopify_import_products.import_all_products",
			});
		}

		// sync progress
		this.logSync();
	}

	logSync() {
		const _log = $("#shopify-sync-log");
		_log.parents(".card").show();
		_log.text(""); // clear logs

		// define counters here to prevent calling jquery every time
		const _syncedCounter = $("#count-products-synced");
		const _erpnextCounter = $("#count-products-erpnext");

		frappe.realtime.on(
			"shopify.key.sync.all.products",
			({ message, synced, done, error }) => {
				message = `<pre class="mb-0">${message}</pre>`;
				_log.append(message);
				_log.scrollTop(_log[0].scrollHeight);

				if (synced)
					this.updateSyncedCount(_syncedCounter, _erpnextCounter);

				if (done) {
					frappe.realtime.off("shopify.key.sync.all.products");
					this.toggleSyncAllButton(false);
					this.fetchProductCount();
					this.syncRunning = false;
				}
			}
		);
	}

	toggleSyncAllButton(disable = true) {
		const btn = $("#btn-sync-all");

		const _toggleClass = (d) => (d ? "btn-success" : "btn-primary");
		const _toggleText = () => (disable ? "Syncing..." : "Sync Products");

		btn.prop("disabled", disable)
			.addClass(_toggleClass(disable))
			.removeClass(_toggleClass(!disable))
			.text(_toggleText());
	}

	updateSyncedCount(_syncedCounter, _erpnextCounter) {
		let _synced = parseFloat(_syncedCounter.text());
		let _erpnext = parseFloat(_erpnextCounter.text());

		_syncedCounter.text(_synced + 1);
		_erpnextCounter.text(_erpnext + 1);
	}
};

File: ecommerce_integrations/shopify/page/shopify_import_products/shopify_import_products.py
from time import process_time

import frappe
from frappe.exceptions import UniqueValidationError
from shopify.resources import Product

from ecommerce_integrations.ecommerce_integrations.doctype.ecommerce_item import ecommerce_item
from ecommerce_integrations.shopify.connection import temp_shopify_session
from ecommerce_integrations.shopify.constants import MODULE_NAME
from ecommerce_integrations.shopify.product import ShopifyProduct
from ecommerce_integrations.shopify.utils import get_user_company, get_user_shopify_account

# constants
SYNC_JOB_NAME = "shopify.job.sync.all.products"
REALTIME_KEY = "shopify.key.sync.all.products"


@frappe.whitelist()
def get_shopify_products(from_=None):
	shopify_products = fetch_all_products(from_)
	return shopify_products


def fetch_all_products(from_=None):
	# format shopify collection for datatable

	collection = _fetch_products_from_shopify(from_)

	products = []
	for product in collection:
		d = product.to_dict()
		d["synced"] = is_synced(product.id)
		products.append(d)

	next_url = None
	if collection.has_next_page():
		next_url = collection.next_page_url

	prev_url = None
	if collection.has_previous_page():
		prev_url = collection.previous_page_url

	return {
		"products": products,
		"nextUrl": next_url,
		"prevUrl": prev_url,
	}


@temp_shopify_session(shopify_account=None)
def _fetch_products_from_shopify(from_=None, limit=20):
	if from_:
		collection = Product.find(from_=from_)
	else:
		collection = Product.find(limit=limit)

	return collection


@frappe.whitelist()
def get_product_count(shopify_account=None):
	items = frappe.db.get_list("Item", {"variant_of": ["is", "not set"]})
	erpnext_count = len(items)

	sync_items = frappe.db.get_list("Ecommerce Item", {"variant_of": ["is", "not set"]})
	synced_count = len(sync_items)

	shopify_count = get_shopify_product_count(shopify_account=shopify_account)

	return {
		"shopifyCount": shopify_count,
		"syncedCount": synced_count,
		"erpnextCount": erpnext_count,
	}


@temp_shopify_session(shopify_account=None)
def get_shopify_product_count():
	return Product.count()


@frappe.whitelist()
def sync_product(product):
	try:
		shopify_product = ShopifyProduct(product, company=get_user_company(frappe.session.user))
		shopify_product.sync_product()

		return True
	except Exception:
		frappe.db.rollback()
		return False


@frappe.whitelist()
def resync_product(product):
	return _resync_product(product)


@temp_shopify_session(shopify_account=None)
def _resync_product(product):
	savepoint = "shopify_resync_product"
	try:
		item = Product.find(product)

		frappe.db.savepoint(savepoint)
		for variant in item.variants:
			shopify_product = ShopifyProduct(product, variant_id=variant.id, company=get_user_company(frappe.session.user))
			shopify_product.sync_product()

		return True
	except Exception:
		frappe.db.rollback(save_point=savepoint)
		return False


def is_synced(product):
	return ecommerce_item.is_synced(MODULE_NAME, integration_item_code=product)


@frappe.whitelist()
def import_all_products():
	shopify_account = get_user_shopify_account()
	if not shopify_account:
		frappe.throw(
			"No Shopify account is linked to your user/company. Please configure Shopify Account and User Permissions before importing products."
		)

	frappe.enqueue(
		queue_sync_all_products,
		queue="long",
		job_name=SYNC_JOB_NAME,
		key=REALTIME_KEY,
		user=frappe.session.user,
		shopify_account=shopify_account.name,
	)


def queue_sync_all_products(*args, **kwargs):
	start_time = process_time()
	shopify_account = kwargs.get("shopify_account")
	synced_count = 0
	failed_count = 0
	failed_products = []

	counts = get_product_count(shopify_account=shopify_account)
	publish("Syncing all products...")

	if counts["shopifyCount"] < counts["syncedCount"]:
		publish("⚠ Shopify has less products than ERPNext.")

	_sync = True
	collection = _fetch_products_from_shopify(limit=100, shopify_account=shopify_account)
	savepoint = "shopify_product_sync"
	while _sync:
		for product in collection:
			try:
				publish(f"Syncing product {product.id}", br=False)
				frappe.db.savepoint(savepoint)
				if is_synced(product.id):
					publish(f"Product {product.id} already synced. Skipping...")
					continue

				shopify_product = ShopifyProduct(product.id, company=get_user_company(kwargs.get("user")))
				shopify_product.sync_product()
				synced_count += 1

				publish(f"✅ Synced Product {product.id}", synced=True)

			except UniqueValidationError as e:
				failed_count += 1
				failed_products.append(str(product.id))
				publish(f"❌ Error Syncing Product {product.id} : {e!s}", error=True)
				frappe.db.rollback(save_point=savepoint)
				continue

			except Exception as e:
				failed_count += 1
				failed_products.append(str(product.id))
				publish(f"❌ Error Syncing Product {product.id} : {e!s}", error=True)
				frappe.db.rollback(save_point=savepoint)
				continue

		if collection.has_next_page():
			frappe.db.commit()  # prevents too many write request error
			collection = _fetch_products_from_shopify(
				from_=collection.next_page_url, shopify_account=shopify_account
			)
		else:
			_sync = False

	end_time = process_time()
	if failed_count and not synced_count:
		publish(
			f"⚠ Sync finished with errors. Synced: {synced_count}, Failed: {failed_count}.",
			error=True,
			done=True,
		)
	elif failed_count:
		publish(
			f"⚠ Sync completed with partial errors. Synced: {synced_count}, Failed: {failed_count}.",
			error=True,
			done=True,
		)
	else:
		publish(f"🎉 Done in {end_time - start_time}s. Synced: {synced_count}", done=True)
	return True


def publish(message, synced=False, error=False, done=False, br=True):
	frappe.publish_realtime(
		REALTIME_KEY,
		{
			"synced": synced,
			"error": error,
			"message": message + ("<br /><br />" if br else ""),
			"done": done,
		},
	)
